### ISSUE TITLE
Weird looking pred +1 and -1

### ISSUE THREAD

**Issue by didriksg** [nianticlabs/monodepth2#278]
First of all, I want to thank you for your amazing contribution!

I have tried to train the system on the Waymo dataset. The images in this dataset consist of sequences of approximately 200, collected from multiple vehicles. Therefore, there are multiple intrinsic parameters associated with different sequences. I have tried to handle this in the dataset class I wrote, but I have a suspicion that I am doing something wrong. 

After running the training steps for 19 epochs, the validation and train loss do have a decreasing trend. However, the numeric difference between the start and end does not seem to have a significant decrease (although I am not quite sure which values to expect here). 

![image](https://user-images.githubusercontent.com/17567687/100226830-91697a80-2f20-11eb-8b7a-65ac3475ec1a.png)

The final predicted +1 and -1 image have some weird looking tendencies, having some kind of wave pattern in some of the predictions:

Example:
![image](https://user-images.githubusercontent.com/17567687/100227052-d68dac80-2f20-11eb-83c5-79210b636dbf.png)

The resulting disparity (from tensorboard)
![image](https://user-images.githubusercontent.com/17567687/100227375-59af0280-2f21-11eb-8d06-c7e55e840efd.png)

Other images look at first sight good (Although with some artifacts on the sides of the image. These might be normal (?) )

pred +1:
![image](https://user-images.githubusercontent.com/17567687/100227591-a5fa4280-2f21-11eb-940d-eecfaa3dbf17.png)

disparity:
![image](https://user-images.githubusercontent.com/17567687/100227615-af83aa80-2f21-11eb-880f-820d7b18d2d2.png)

This is my dataset class:

```
class WaymoDataset(MonoDataset):
    def __init__(self, *args, **kwargs):
        super(WaymoDataset, self).__init__(*args, **kwargs)
        self.full_res_shape = (1920, 1280)

    def check_depth(self):
        return False

    def get_depth(self, folder, frame_index, side, do_flip):
        raise NotImplementedError

    def get_image_path(self, folder, frame_index):
        path = f'{self.data_path}{folder}/img_resized_{frame_index+1}.jpg'

        return path

    def get_color(self, folder, frame_index, side, do_flip):
        color = self.loader(self.get_image_path(folder, frame_index))

        if do_flip:
            color = color.transpose(pil.FLIP_LEFT_RIGHT)

        return color

    def get_k_matrix(self, folder):
        K = np.load(f'{self.data_path}{folder}/k_matrix.npy')

        K[0, 0] /= self.full_res_shape[0]
        K[0, 2] /= self.full_res_shape[0]
        K[1, 1] /= self.full_res_shape[1]
        K[1, 2] /= self.full_res_shape[1]

        return K
```

The `get_k_matrix` function replaces line 165 in `mono_dataset.py`

`        for scale in range(self.num_scales):
            K = self.get_k_matrix(folder)
`

My split looks like this:
- training_0005/segment-11252086830380107152_1540_000_1560_000 0
- ...
- training_0005/segment-11252086830380107152_1540_000_1560_000 197

The index is 0 indexed, and contain all files in the folder, e.g. the segment folder contains image_0 --> image_197. I figured that the dataloader fixes the problem with the final file, and the indexing of getting image `frame_index+1` fixes the problem of requesting image_-1.

An example K matrix looks like this:
[[2.051289e+03 0.000000e+00 9.505541e+02 0.000000e+00]
 [0.000000e+00 2.051289e+03 6.471451e+02 0.000000e+00]
 [0.000000e+00 0.000000e+00 1.000000e+00 0.000000e+00]
 [0.000000e+00 0.000000e+00 0.000000e+00 1.000000e+00]]

The original image height and width are 1920x1280. Although, I have resized the images used in training to be 480x320, keeping the original aspect ratio and the requirement of a height divisible by 32.

Any suggestion or point in the right direction is highly appreciated!

**Comment by mrharicot** [comment]
Hi,

I am not seeing anything suspicious so far in your code, however we do not handle the first and last images of a sequence and simply never use them as "middle frame" when training.
Specifically the split in the text file should look like this:

- training_0005/segment-11252086830380107152_1540_000_1560_000 1
- training_0005/segment-11252086830380107152_1540_000_1560_000 2
- ...
- training_0005/segment-11252086830380107152_1540_000_1560_000 195
- training_0005/segment-11252086830380107152_1540_000_1560_000 196

Another possibility is whether the images in the waymo dataset have been corrected for distortion.
monodepth2 only supports linear cameras.

I hope this helps!

**Comment by didriksg** [comment]
Hey!

Thanks for your suggestions

I will try to update my splits. I have tried to look in a bit more detail in the dataset and I have noticed that there are some occurrences of the vehicle collecting the data standing still and not moving. I have also noticed that this dataset most likely has a higher FPS than the Kitti dataset. I can try to skip every other frame or so, and see whether or not it has any effect. 

I also tried to train with the Kitti data, using the eigen_zhou split. I noticed that the initial loss is much higher (and also has a more expectable loss development). The Waymo dataset's initial loss values are pretty consistently around the 0.04-0.06 range and do not develop very much during training. Could this be a sign of too little change between frames? The Waymo dataset also has quite a few sections of images captured during rain (with accompanying droplets in the image) and during low lighting situations, where images tend to be more pixelated, like the one I showed earlier. Idk how much this changes things tho.. Just trying to write down all anomalies I can think of.

I will test a bit more and let you know if I get any results!

**Comment by didriksg** [comment]
Here is btw some other examples of +1 and -1s after 8000 iterations:
![image](https://user-images.githubusercontent.com/17567687/100683976-ba9f6600-3379-11eb-93bf-5ad5eed2c6e2.png)
![image](https://user-images.githubusercontent.com/17567687/100683992-c1c67400-3379-11eb-9865-ce63b6db0ca4.png)
![image](https://user-images.githubusercontent.com/17567687/100684014-cbe87280-3379-11eb-9fbc-2d81ffbd18e8.png)
![image](https://user-images.githubusercontent.com/17567687/100684026-d0ad2680-3379-11eb-9a3e-7bfbea88e13d.png)

**Comment by mrharicot** [comment]
Note that you can change the reference frames in the command line you run to train: `--frame_ids 0 -2 2` for instance.
Also have a look at an older post regarding the waymo dataset: https://github.com/nianticlabs/monodepth2/issues/278
I believe there was at least another issue on the waymo dataset but I can't seem to be able to find it.

**Comment by mrharicot** [comment]
Are you training with all the cameras? I would suggest using only the forward camera for now to simplify your debugging.

**Comment by didriksg** [comment]
I'm only using the forward camera atm, yes.

**Comment by didriksg** [comment]
I've looked into isse #160 , but the problem there was the Intrinsic parameters. I did also notice that he used a static K matrix tho. Although as these changes from segment to segment, my initial thought was that doing so would cause problems by itself.

**Comment by mrharicot** [comment]
I would also consider using a lower resolution training resolution to regularize the results and train faster.

**Comment by didriksg** [comment]
How low can I go before results stops making sense? I'm currently using 480x320

**Comment by mdfirman** [comment]
I agree with @mrharicot here – training lower resolution should help to get you started. Maybe try 240 x 160 even.

Using forward camera only sounds helpful.

I agree that getting the fps/sampling rate right is important. I would try to get a frame sampling roughly the same as KITTI. You could flick through some sequences to try to judge that sampling rate by eye.

If there are too many static images (where the camera isn't moving) this could cause a problem too. Some people remove these types of images from their datasets.

Finally, you could try disabling full resolution multiscale (with the `--v1_multiscale`). It might not help to disable it, but it might.

**Comment by mrharicot** [comment]
Closing this for now.
